### ISSUE TITLE
gateway-proxy returning 503 after rotating certificate in upstream

### ISSUE THREAD

**Issue by jmunozro** [solo-io/gloo#6337]
### Gloo Edge Version

1.11.x (latest stable)

### Kubernetes Version

1.23.x

### Describe the bug

gateway-proxy is not able to handle a certificate rotation for an upstream

### Steps to reproduce the bug

1. install gloo with some common options
```
export GLOO_VERSION=1.11.3
helm repo update
helm upgrade --install gloo glooe/gloo-ee --namespace gloo-system \
  --create-namespace --version ${GLOO_VERSION} -f - <<EOF
license_key: $LICENSE_KEY
gloo:
  gateway:
    compressedProxySpec: true
  gatewayProxies:
    gatewayProxy:
      kind:
        deployment:
          replicas: 1
      antiAffinity: true
      podTemplate:    # Workaround from Gloo developers - https://github.com/solo-io/gloo/issues/5494
        probes: true # not needed for shutdown
        gracefulShutdown:
          enabled: true # force the healthcheck/fail, then wait 5s, and then start shutdown
          sleepTimeSeconds: 5 # wait 5s to start shutdown       
      gatewaySettings:
        accessLoggingService:
          accessLog:
          - fileSink:
              path: /dev/stdout
              stringFormat: >
                [%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"
                %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION%
                %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%" "%REQ(X-OSTenant)%"
                "%REQ(X-REQUEST-ID)%" "%REQ(x-amz-cf-id)%" "%REQ(:AUTHORITY)%" "%UPSTREAM_HOST%" "%DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT%"
  #discovery:
  #  enabled: false
  settings:
    replaceInvalidRoutes: true
    invalidConfigPolicy:
      invalidRouteResponseBody: Gloo Edge has invalid configuration. Contact administrators to find and fix config errors (glooctl check).
      invalidRouteResponseCode: 404
      replaceInvalidRoutes: true
global:
  extensions:         
    extAuth:
      enabled: true        
      envoySidecar: true
      standaloneDeployment: false
EOF
```

2. Deploy Keycloak
```
kubectl create -f https://gist.githubusercontent.com/jmunozro/a26a3b91922b503c1d3bf7371fbb263e/raw/682f6189e1b1f9a43fb3cecff088a4e336a5d447/keycloak.yaml
```
3. Create Root CA
```
step certificate create root-ca-edge root-ca.crt root-ca.key \
  --profile root-ca --no-password --insecure --not-after 87600h --kty RSA
```

4. Create httpbin certificate
```
step certificate create httpbin.solo.io tls.crt tls.key --ca root-ca.crt \
    --ca-key root-ca.key --not-after 8760h --no-password --insecure \
    --san httpbin.solo.io --kty RSA
```

5. Create upstream tls
```
kubectl create secret tls upstream-tls --key tls.key \
   --cert tls.crt --namespace gloo-system
```

6. Create identity-api-ssl with the same cert
```
kubectl create secret tls identity-api-ssl --key tls.key \
   --cert tls.crt --namespace gloo-system

kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: keycloak
  namespace: gloo-system
spec:
  sslConfig:
    secretRef:
      name: upstream-tls
      namespace: gloo-system
  virtualHost:
    domains:
    - httpbin.solo.io
    routes:
    - matchers:
      - prefix: /
      routeAction:
        single:
          upstream:
            name: keycloak-tls-server
            namespace: gloo-system
---
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: keycloak-tls-server
  namespace: gloo-system
spec:
  kube:
    selector:
      app: keycloak
    serviceName: keycloak
    serviceNamespace: default
    servicePort: 8443
  sslConfig:
    secretRef:
      name: identity-api-ssl
      namespace: gloo-system
EOF
```

7. It works
```
curl https://httpbin.solo.io/auth/ --resolve httpbin.solo.io:443:35.204.148.47 -isk
```

9. Create httpbin certificate
```
step certificate create httpbin.solo.io tls.crt tls.key --ca root-ca.crt \
    --ca-key root-ca.key --not-after 8760h --no-password --insecure \
    --san httpbin.solo.io --kty RSA

kubectl create secret tls identity-api-ssl --key tls.key \
   --cert tls.crt --namespace gloo-system
```

10. Rotate the certificate, use the same root CA
```
step certificate create httpbin.solo.io tls.crt tls.key --ca root-ca.crt \
    --ca-key root-ca.key --not-after 8760h --no-password --insecure \
    --san httpbin.solo.io --kty RSA -f
kubectl create secret tls identity-api-ssl \
--save-config \
--dry-run=client \
--key tls.key \
--cert tls.crt --namespace gloo-system \
-o yaml | \
kubectl apply -f -
```

11. It fails, fix it
```
curl https://httpbin.solo.io/auth/ --resolve httpbin.solo.io:443:35.204.148.47 -isk
k -n gloo-system rollout restart deployment gateway-proxy
```

in the logs we see
`gRPC config: initial fetch timed out for [type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment](http://type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment)`, which can be also seen in the stats of the video below

### Expected Behavior

Sometimes the certificate is rotated and you can see the stats inmediately restarted. Other times they take some time and the cluster become unhealthy.

For simplicity, in the example I used a single gateway-proxy replica, and a single keycloak replica, but with more gateway-proxy replicas the chances of SOME of them returning 503 is high. What I found in the customer is that with 3 replicas usually 1 of them had the issue. They rotate the certificate with cert-manager periodically.

It can be fixed with a bounce, but it is unpredictable and dangerous in production.

### Additional Context

TL;DR
It breaks after the first minute of the video, about 1:18.

https://user-images.githubusercontent.com/35881711/164031143-b0ea533d-b5d5-454d-b11f-cdc3ca50aab4.mov

**Comment by jmunozro** [comment]
I just discovered that making any change to the upstream (adding a label) seems to fix it.
```
kubectl apply -f - <<EOF
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: keycloak-tls-server
  namespace: gloo-system
  labels:
    foo: bar
spec:
  kube:
    selector:
      app: keycloak
    serviceName: keycloak
    serviceNamespace: default
    servicePort: 8443
  sslConfig:
    secretRef:
      name: identity-api-ssl
      namespace: gloo-system
EOF
```

**Comment by jackstine** [comment]
Will be on the next release for v12 and v11

**Comment by jackstine** [comment]
There was a slight error in the configuration, draft is up now for review.

**Comment by jackstine** [comment]
This is in Gloo v1.11.16
Gloo-EE v1.11.23
